[triciiaaa] iP

triciiaaa opened the pull request:
# ⏲️ Chronos

> Your Comprehensive Time Management Companion for Effortless Organisation and Productivity.

## 🔥 Benefits  
- Efficient Time Management
- Improved Productivity
- Enhanced Organisation

## 🚀  Installation and Usage
1. Download the latest release of chronos.jar from [here](https://github.com/triciiaaa/ip/releases).
2. Copy the file to the folder you want to use as the _home folder_ for your Chronos application.
3. Open a command terminal, `cd` into the folder you put the jar file in, and use the `java -jar chronos.jar` command to run the application.

## ⚙️ Features
**Types of Tasks**: Todos, Deadlines, Events
- [x] Managing Tasks (View, Search, Add, Delete, Mark, Unmark)
- [ ] Reminders (coming soon)

## 🌟 `main` method
``` java
public static void main(String[] args) throws IOException, ChronosException {
     new Chronos(FILE_PATH).runProgram();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

triciiaaa committed:
Level-0 Increment

triciiaaa committed:
Level-1 Increment + Added styling with lines and indentation

triciiaaa committed:
Level-2 Increment

triciiaaa committed:
Level-3 Increment

triciiaaa committed:
Level-4 Increment + Inheritance of Task

triciiaaa committed:
Level-4 Increment + Inheritance of Task

triciiaaa committed:
A-TextUiTesting Increment + Resolved Bugs

triciiaaa committed:
Level-5 Increment + Error Handling for Existing Features

triciiaaa committed:
Level-6 Increment

triciiaaa committed:
Level 7 Increment

triciiaaa committed:
Amend code structure and format

triciiaaa committed:
Add Java documentation for existing code

triciiaaa committed:
Merge branch 'branch-Level-7'

triciiaaa committed:
Amend code structure and format

triciiaaa committed:
Amend date validation for Deadline objects

triciiaaa committed:
Amend date and time validation for Deadline and Event objects

triciiaaa committed:
Amend Java documentation for existing code

triciiaaa committed:
Merge branch 'branch-Level-8'

triciiaaa committed:
Create custom Chronos Exception class and sub-classes

kaiyi27 commented on src/main/java/Chronos.java:
```diff
@@ -0,0 +1,355 @@
+import java.util.ArrayList;
+import java.util.Scanner;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
```
Are these import statement following the right order?
Maybe consider follow by the order of lexicographical.

kaiyi27 commented on src/main/java/Chronos.java:
```diff
@@ -0,0 +1,355 @@
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+
+import task.Task;
+import task.Todo;
+import task.Deadline;
+import task.Event;
```
Should these classes being packaged into a single package to use it more conveniently?

kaiyi27 commented on src/main/java/Chronos.java:
```diff
@@ -0,0 +1,355 @@
+ * Represents the main class of the Chronos Task Management System.
+ */
+public class Chronos {
+    private static final String DIVIDER = "        ------------------------------------------------------------";
+    private static final String POSSIBLE_COMMANDS = "        TODO     --- todo [task name]\n" +
+                                                    "        DEADLINE --- deadline [task name] /by [yyyy-mm-dd HH:MM]\n" +
+                                                    "        EVENT    --- event [task name] /from [yyyy-mm-dd HH:MM] /to [yyyy-mm-dd HH:MM]" ;
```
Maybe consider break before an operator.
e.g: a + b
       + c

kaiyi27 commented on src/main/java/Chronos.java:
```diff
@@ -0,0 +1,355 @@
+                                                    "        DEADLINE --- deadline [task name] /by [yyyy-mm-dd HH:MM]\n" +
+                                                    "        EVENT    --- event [task name] /from [yyyy-mm-dd HH:MM] /to [yyyy-mm-dd HH:MM]" ;
+
+    private static String filePath = "./data/chronos.txt";
```
Is it possible to make the filepath final value?

kaiyi27 commented on src/main/java/Chronos.java:
```diff
@@ -0,0 +1,355 @@
+
+    private static String filePath = "./data/chronos.txt";
+    private static ArrayList<Task> tasks = new ArrayList<Task>();
+    private static int noOfTasks = 0;
```
Is this noOfTasks value necessary to use? As you may keep track of the noOfTasks by just calling tasks.size();

kaiyi27 commented on src/main/java/Chronos.java:
```diff
@@ -0,0 +1,355 @@
+                            String[] descriptionAndDate = token[1].split("/from");
+                            String description = descriptionAndDate[0].trim();
+                            String[] fromAndTo = descriptionAndDate[1].split("/to");
+                            String fromDateAndTime = Chronos.formatDateTime(fromAndTo[0].trim());
+                            String toDateAndTime = Chronos.formatDateTime(fromAndTo[1].trim());
+                            Chronos.addEvent(description, fromDateAndTime, toDateAndTime);
+                            Chronos.saveTasks(fw);
+                        } catch (Exception e) {
+                            System.out.println(DIVIDER);
+                            System.out.println("        Invalid command. Please include a task name and a valid due date following the syntax of the example below:");
+                            System.out.println("        e.g. event concert /from 2024-02-16 18:00 /to 2024-02-16 20:00");
+                            System.out.println(DIVIDER);
+                        }
+                        break;
+                    case "delete":
+                        if (token.length != 2 || token[1].trim().isEmpty()) {
+                            System.out.println(DIVIDER);
+                            System.out.println("        Missing task number. Please include the task number to be deleted.");
+                            System.out.println(DIVIDER);
+                        } else {
+                            try {
+                                int i = Integer.parseInt(token[1]);
+                                Chronos.deleteTask(i);
+                                Chronos.saveTasks(fw);
+                            } catch (NumberFormatException e) {
+                                System.out.println(DIVIDER);
+                                System.out.println("        Task number is not an integer. Please include a valid task number.");
+                                System.out.println(DIVIDER);
+                            } catch (IndexOutOfBoundsException e) {
+                                System.out.println(DIVIDER);
+                                System.out.println("        Task number out of range. Please include a valid task number.");
+                                System.out.println(DIVIDER);
+                            }
+                        }
+                        break;
+                    default:
+                        System.out.println(DIVIDER);
+                        System.out.println("        Invalid command. Please try again.");
+                        System.out.println(DIVIDER);
+                        break;
```
Should there be indention for the case and default statement?

kaiyi27 reviewed (commented):
Overall, the code looks good to me, I found your code easy to read for the most part except a few places where the nesting was too deep. I noted a few minor coding standard violations too. The main class are getting quite long. Consider splitting into smaller classes and method if that makes sense.

triciiaaa committed:
Refactor code to additional classes (Ui, Storage, Parser and TaskList)

triciiaaa committed:
Fix load tasks from file feature

triciiaaa committed:
Add Java documentation for Exception classes

triciiaaa committed:
Add Java documentation for Task classes

triciiaaa committed:
Add Java documentation for Parser, Storage, TaskList and Ui classes

triciiaaa committed:
Refactor code in Ui and TaskList classes

triciiaaa committed:
Merge branch 'A-MoreOOP'

triciiaaa committed:
Divide classes into packages

triciiaaa committed:
Merge branch 'A-Packages'

triciiaaa committed:
Merge branch 'add-gradle-support'

GiselleTan commented on src/main/java/tool/TaskList.java:
```diff
@@ -0,0 +1,129 @@
+     *
+     * @return Number of tasks in the list.
+     */
+    public int size() {
```
perhaps naming it as getSize() would make it clearer when this function is used that it is returning the size of the task list?

GiselleTan commented on src/main/java/tool/Ui.java:
```diff
@@ -0,0 +1,190 @@
+     */
+    public void greetUser() {
+        System.out.println(DIVIDER);
+        System.out.println("        Hello! I'm Chronos.");
+        System.out.println("        What can I do for you?");
```
I noticed that you're writing out the spaces yourself in several parts of the code, perhaps using "\t" would be easier as it adds a tab space?

GiselleTan commented on src/main/java/tool/Ui.java:
```diff
@@ -0,0 +1,190 @@
+    /**
+     * Prints divider.
+     */
+    public void showDivider() {
```
is this function being used?

GiselleTan reviewed (commented):
Overall, I loved reading your code. It was really easy to read, understand and well-organised.

triciiaaa committed:
Add unit tests for two methods in Parser and Storage classes

triciiaaa committed:
Merge branch 'A-JUnit'

triciiaaa committed:
Edit build.gradle file

triciiaaa committed:
Check previously added Java documentation

triciiaaa committed:
Adhere to Java Coding Standard (Naming)

triciiaaa committed:
Adhere to Java Coding Standard (Layout)

triciiaaa committed:
Adhere to Java Coding Standard (Statements)

triciiaaa committed:
Adhere to Java Coding Standard (Comments)

triciiaaa committed:
Merge branch 'A-JavaDoc'

triciiaaa committed:
Merge branch 'A-CodingStandard'

triciiaaa committed:
Add 'find' command to find task(s) through keyword(s)

triciiaaa committed:
Merge branch 'branch-Level-9'

triciiaaa committed:
Use checkStyle to detect any coding style violations

triciiaaa committed:
Replace manual tabs with "\t\t" and fix minor bugs

triciiaaa committed:
Merge branch 'branch-A-CheckStyle'

triciiaaa committed:
Change Text UI to GUI using JavaFX

triciiaaa committed:
Merge branch 'branch-Level-10'

triciiaaa committed:
Configure build.gradle file

The main class has changed, and it will affect the referencing on the
variable mainClassName and archiveFileName.

Changing the value of mainClassName to Launcher and the value of
archiveFileName to 'chronos.jar' allows us to create a fat JAR file
containing the JavaFX libraries. This is essential for the GUI
implemented for the Chronos application.

triciiaaa committed:
Add Java's assert feature

It serves as a tool to document important assumptions that should hold
at various points of the code.

Adding assertions in Parser.java helps to check whether the task list
is not empty upon adding a todo, deadline or event, as well as, to check
whether the retrieved task is not null using the getTask() method.

triciiaaa committed:
Add Java documentation

triciiaaa committed:
Adhere to Code Quality

Long methods are abstracted into smaller methods to improve
readability.

triciiaaa committed:
Merge pull request #2 from triciiaaa/branch-A-Assertions

Add Java's assert feature

triciiaaa committed:
Merge branch 'master' of https://github.com/triciiaaa/ip

triciiaaa committed:
Merge pull request #3 from triciiaaa/branch-A-CodeQuality

Adhere to Code Quality

triciiaaa committed:
Merge branch 'master' of https://github.com/triciiaaa/ip

triciiaaa committed:
Add gradle.yml file

triciiaaa committed:
Merge branch 'branch-A-CI'

triciiaaa committed:
Fix gradle build errors

triciiaaa committed:
Implement feature extension

Making an alternative shorter command syntax helps to increase the
efficiency of the user when using the application.

triciiaaa committed:
Merge branch 'branch-C-FriendlierSyntax'

triciiaaa committed:
Improve GUI

Changing text box colour makes it easier to differentiate between
Chronos text and user text. Truncated message bug is fixed too.

triciiaaa committed:
Merge branch 'branch-A-BetterGui'

triciiaaa committed:
Fix gradle build errors

triciiaaa committed:
Update README.md

triciiaaa committed:
Update README.md

triciiaaa committed:
Update README.md

triciiaaa committed:
Update README.md

triciiaaa committed:
Update README.md

triciiaaa committed:
Update README.md

triciiaaa committed:
Update README.md

triciiaaa committed:
Set custom colours and allow program to terminate upon bye command

triciiaaa committed:
Fix gradle build error

triciiaaa committed:
Clean up methods to adhere to code quality standards

triciiaaa committed:
Update README.md

triciiaaa committed:
Update README.md

triciiaaa committed:
Update README.md

triciiaaa committed:
Reduce size of profile image

triciiaaa committed:
Update README.md

triciiaaa committed:
Merge branch 'master' of https://github.com/triciiaaa/ip
